pkg/frontend/querymiddleware/query_limiter.go

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package querymiddleware
+
+import (
+	"context"
+	"errors"
+	"strings"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/cache"
+	"github.com/grafana/dskit/tenant"
+	"github.com/prometheus/client_golang/prometheus"
+
+	"github.com/grafana/mimir/pkg/util/spanlogger"
+	"github.com/grafana/mimir/pkg/util/validation"
+)
+
+// queryLimiterMiddleware blocks a query if it is should not be executed more frequently than a configured interval,
+// and allows it if enough time has passed since the last time the query was allowed.
+// When blocked, the request is rejected with a "too many requests" error.
+// The query limiter currently only matches exact queries, and does not check against query time, only execution time.
+type queryLimiterMiddleware struct {
+	next                  MetricsQueryHandler
+	cache                 cache.Cache
+	keyGen                CacheKeyGenerator
+	limits                Limits
+	logger                log.Logger
+	blockedQueriesCounter *prometheus.CounterVec
+}
+
+func newQueryLimiterMiddleware(
+	cache cache.Cache,
+	cacheKeyGen CacheKeyGenerator,
+	limits Limits,
+	logger log.Logger,
+	blockedQueriesCounter *prometheus.CounterVec,
+) MetricsQueryMiddleware {
+	return MetricsQueryMiddlewareFunc(func(next MetricsQueryHandler) MetricsQueryHandler {
+		return &queryLimiterMiddleware{
+			next:                  next,
+			cache:                 cache,
+			keyGen:                cacheKeyGen,
+			limits:                limits,
+			logger:                logger,
+			blockedQueriesCounter: blockedQueriesCounter,
+		}
+	})
+}
+
+func (ql *queryLimiterMiddleware) Do(ctx context.Context, req MetricsQueryRequest) (Response, error) {
+	spanLog := spanlogger.FromContext(ctx, ql.logger)
+	tenantIDs, err := tenant.TenantIDs(ctx)
+	if err != nil {
+		return ql.next.Do(ctx, req)
+	}
+
+	key := ql.keyGen.QueryRequestLimiter(ctx, tenant.JoinTenantIDs(tenantIDs), req)
+	hashedKey := cacheHashKey(key)
```

Comment:
The hash being done here is not a cryptographic hash so collisions are possible. We work around that elsewhere in the query-frontend by also storing the entire key as part of the cache payload. Since we're not doing that here, I think we need to use a cryptographic hash for this key. There are a few examples of using the `blake2b` hash in Mimir:

* [CachingBucket](https://github.com/grafana/mimir/blob/1eef2a6e0e67a5f767630b5eec08dea173f14dc1/pkg/storage/tsdb/bucketcache/caching_bucket.go#L799-L809)
* [IndexCache](https://github.com/grafana/mimir/blob/1eef2a6e0e67a5f767630b5eec08dea173f14dc1/pkg/storegateway/indexcache/remote.go#L141-L187)